File: backend/src/models/Bill.js
const mongoose = require('mongoose');

const billItemSchema = new mongoose.Schema({
  name: {
    type: String,
    required: true
  },
  quantity: {
    type: Number,
    required: true,
    default: 1
  },
  cost: {
    type: Number,
    required: true
  }
}, { _id: true });

const itemAssignmentSchema = new mongoose.Schema({
  itemId: {
    type: mongoose.Schema.Types.ObjectId,
    required: true
  },
  participant: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  quantity: {
    type: Number,
    required: true,
    default: 1
  },
  claimedAt: {
    type: Date,
    default: Date.now
  }
}, { _id: true });

const billSchema = new mongoose.Schema({
  uploadedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  imageUrl: {
    type: String,
    required: true
  },
  participants: [{
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  }],
  items: [billItemSchema],
  totalAmount: {
    type: Number,
    default: 0
  },
  status: {
    type: String,
    enum: ['uploaded', 'processing', 'processed', 'split', 'finalized'],
    default: 'uploaded'
  },
  assignmentMode: {
    type: String,
    enum: ['not_set', 'uploader_assigns', 'self_select'],
    default: 'not_set'
  },
  itemAssignments: [itemAssignmentSchema],
  shares: [{
    participant: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    amount: {
      type: Number,
      default: 0
    }
  }],
  participantsFinished: [{
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  }],
  restaurant: {
    name: String,
    type: {
      type: String,
      enum: ['restaurant', 'bar'],
      default: 'restaurant'
    }
  },
  createdAt: {
    type: Date,
    default: Date.now
  }
}, {
  timestamps: true
});

// Calculate total amount from items
billSchema.methods.calculateTotal = function() {
  this.totalAmount = this.items.reduce((sum, item) => {
    return sum + (item.cost * item.quantity);
  }, 0);
  return this.totalAmount;
};

// Calculate shares based on item assignments
billSchema.methods.calculateShares = function() {
  const sharesMap = new Map();

  // Initialize shares for all participants
  this.participants.forEach(participantId => {
    sharesMap.set(participantId.toString(), 0);
  });

  // Calculate shares based on assignments
  this.itemAssignments.forEach(assignment => {
    const item = this.items.id(assignment.itemId);
    if (item) {
      const participantId = assignment.participant.toString();
      const itemTotal = item.cost * assignment.quantity;
      const currentAmount = sharesMap.get(participantId) || 0;
      sharesMap.set(participantId, currentAmount + itemTotal);
    }
  });

  // Update shares array
  this.shares = Array.from(sharesMap.entries()).map(([participantId, amount]) => ({
    participant: participantId,
    amount: Math.round(amount * 100) / 100 // Round to 2 decimal places
  }));

  return this.shares;
};

// Check if all items are fully assigned
billSchema.methods.areAllItemsAssigned = function() {
  for (const item of this.items) {
    const assignedQuantity = this.itemAssignments
      .filter(a => a.itemId.toString() === item._id.toString())
      .reduce((sum, a) => sum + a.quantity, 0);

    if (assignedQuantity < item.quantity) {
      return false;
    }
  }
  return true;
};

module.exports = mongoose.model('Bill', billSchema);

File: ios-app-xcode/Splitzy/Splitzy/Models/Bill.swift
import Foundation

struct Bill: Identifiable, Codable {
    let id: String
    let uploadedBy: User
    let imageUrl: String
    var participants: [User]
    var items: [BillItem]
    var totalAmount: Double
    var status: BillStatus
    var assignmentMode: AssignmentMode
    var itemAssignments: [ItemAssignment]
    var shares: [BillShare]
    var restaurant: Restaurant
    let createdAt: String

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case uploadedBy, imageUrl, participants, items, totalAmount, status
        case assignmentMode, itemAssignments, shares, restaurant, createdAt
    }
}

struct BillItem: Identifiable, Codable {
    var id: String?
    var name: String
    var quantity: Int
    var cost: Double

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case name, quantity, cost
    }
}

struct Restaurant: Codable {
    var name: String
    var type: RestaurantType

    enum CodingKeys: String, CodingKey {
        case name
        case type
    }
}

enum RestaurantType: String, Codable {
    case restaurant = "restaurant"
    case bar = "bar"

    var displayName: String {
        rawValue.capitalized
    }
}

enum BillStatus: String, Codable {
    case uploaded = "uploaded"
    case processing = "processing"
    case processed = "processed"
    case split = "split"
    case finalized = "finalized"

    var displayName: String {
        switch self {
        case .uploaded: return "Uploaded"
        case .processing: return "Processing"
        case .processed: return "Processed"
        case .split: return "Split"
        case .finalized: return "Finalized"
        }
    }
}

enum AssignmentMode: String, Codable {
    case notSet = "not_set"
    case uploaderAssigns = "uploader_assigns"
    case selfSelect = "self_select"

    var displayName: String {
        switch self {
        case .notSet: return "Not Set"
        case .uploaderAssigns: return "I'll Assign"
        case .selfSelect: return "Let Them Pick"
        }
    }
}

struct ItemAssignment: Identifiable, Codable {
    var id: String?
    let itemId: String
    let participant: User
    let quantity: Int
    let claimedAt: String

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case itemId, participant, quantity, claimedAt
    }
}

struct BillShare: Identifiable, Codable {
    var id: String { participant.id }
    let participant: User
    let amount: Double
}

// API Response models
struct BillResponse: Codable {
    let success: Bool
    let data: Bill
}

struct BillsResponse: Codable {
    let success: Bool
    let data: [Bill]
}
